synapseclient/table.py

```diff
@@ -780,6 +781,112 @@ def __init__(self, name=None, columns=None, parent=None, definingSQL=None, prope
+    :param parent:          The Synapse Project to which this Dataset belongs
+    :param properties:      A map of Synapse properties
+    :param annotations:     A map of user defined annotations
+    :param datasetItems:    A list of items characterized by entityId and versionNumber
+    :param local_state:     Internal use only
+
+    Example::
+
+        from synapseclient import Dataset
+        # Create a Dataset with pre-defined DatasetItems
+        datasetItems = [
+            {'entityId': "syn000",
+             'versionNumber: 1},
+            {...},
+        ]
+        dataset = syn.store(Dataset(
+            name="My Dataset",
+            parent=project,
+            datasetItems=datasetItems))
+
+        # Add/remove specific Synapse IDs to the Dataset
+        dataset.addItem({'entityId': "syn111", 'versionNumber': 1})
+        dataset.addItems([
+            {'entityId': "syn222", 'versionNumber': 2},
+            {'entityId': "syn333", 'versionNumber': 1}])
+        dataset.removeItem("syn000")
+
+        # Add only top-level Files within a Folder to the Dataset (recursive=False (default))
+        folder_ids = ["syn123", "syn456"]
+        dataset.addFolders(folder_ids, recursive=False)
+
+        # Add all Files within a Folder to the Dataset (recursive=True)
+        dataset.addFolder("syn789", recursive=True)
+
+    To create a snapshot version of the Dataset, use
+    :py:classmethod:`synapseclient.client.create_snapshot_version`.
+
+    Example::
+        syn = synapseclient.login()
+        syn.create_snapshot_version(
+            dataset.id,
+            label="v1.0",
+            comment="This is version 1")
+    """
+    _synapse_entity_type: str = "org.sagebionetworks.repo.model.table.Dataset"
+    _property_keys: List[str] = SchemaBase._property_keys + ['datasetItems']
+
+    def __init__(self, name=None, columns=None, parent=None, properties=None,
+                 annotations=None, local_state=None, datasetItems=None, **kwargs):
+        self.properties.setdefault('datasetItems', [])
+        super(Dataset, self).__init__(
+            name=name, columns=columns, properties=properties,
+            annotations=annotations, local_state=local_state, parent=parent, **kwargs
+        )
+        if datasetItems:
+            self.addItems(datasetItems)
+
+    def addItem(self, dataset_item: Dict[str, str]):
+        """
+        :param dataset_item: a single dataset item
+        """
+        if isinstance(dataset_item, dict):
+            required_keys = {'entityId', 'versionNumber'}
+            if required_keys == dataset_item.keys():
+                self.properties.datasetItems.append(dataset_item)
+            elif required_keys - dataset_item.keys():
+                raise LookupError("DatasetItem missing a required property: %s" %
+                                  str(required_keys - dataset_item.keys()))
+
+            # FIXME? Having other key-value pair in the dict doesn't throw any
```

Comment:
Oh interesting.  So as an item, you can have:
```
{"bogus": "trial"}
```
And the API won't throw an error?

Reply:
I think the API will throw an error if either `entityId` and/or `versionNumber` is missing. But it will not throw an error if there are extra item pairs, e.g.

```python
>>> dataset = syn.get("syn29430164")
>>> dataset.properties.datasetItems.append({'entityId': "syn123", 'versionNumber': 1, 'bogus': "trial"})
>>> dataset.datasetItems
[{'entityId': 'syn20685093', 'versionNumber': 6}, {'entityId': 'syn29461520', 'versionNumber': 1}, {'entityId': 'syn123', 'versionNumber': 1, 'bogus': 'trial'}]
>>> 
>>> 
>>> dataset = syn.store(dataset)
>>> dataset.datasetItems
[{'entityId': 'syn20685093', 'versionNumber': 6}, {'entityId': 'syn29461520', 'versionNumber': 1}, {'entityId': 'syn123', 'versionNumber': 1}]
```

Again, not sure if this case should even be handled, since the extras are not kept after storing.

Reply:
I think that's fine. We can just check for if the two requires keys are there but not for extras